services/spar/src/Spar/Data.hs

```diff
@@ -419,6 +430,47 @@ deleteIdPRawMetadata idp = retry x5 . write del $ params Quorum (Identity idp)
     del :: PrepQuery W (Identity SAML.IdPId) ()
     del = "DELETE FROM idp_raw_metadata WHERE id = ?"
 
+----------------------------------------------------------------------
+-- SSO settings
+
+-- It's important to maintain two invariants:
+-- 1) whenever there is a default code, it must also exist in the idp table
+-- 2) there can always only be one default SSO code selected
+
+getDefaultSSOCode
+  :: (HasCallStack, MonadClient m)
+  => m (Maybe SAML.IdPId)
+getDefaultSSOCode = fmap runIdentity . minimumMay <$>
+  (retry x1 . query sel $ params Quorum ())
+  where
+    sel :: PrepQuery R () (Identity SAML.IdPId)
+    sel = "SELECT idp FROM default_idp WHERE partition_key_always_default = 'default'"
+
+storeDefaultSSOCode
+  :: (HasCallStack, MonadClient m)
+  => SAML.IdPId -> m ()
+storeDefaultSSOCode idpId = do
+  -- there is a race condition here which means there could potentially be more
+  -- than one entry (violating invariant 2).
+  -- However, the SELECT query will deterministally pick one of them and the
+  -- others will get removed by TRUNCATE the next time this function is called.
```

Comment:
where does the `TRUNCATE` happen?  or do you mean `DELETE`?

and is the determinism part of the semantics of cql?  reference the docs then?  (not sure i'm overdoing the referencing here.)

Reply:
Ah, thanks. I wanted to add an `ORDER BY` clause to the `SELECT` query. This will ensure that it picks one deterministically.

Reply:
Done. Better?